### ISSUE TITLE
HttpServerKeepAliveHandler does not account for some self-defined response body lengths

### ISSUE THREAD

**Issue by Bennett-Lynch** [netty/netty#10868]
`HttpServerKeepAliveHandler` has logic to close a connection if a response does not have a self-defined body length:

https://github.com/netty/netty/blob/e8b5aeddef6f342e5e8ca1e03092c7b278da2915/codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java#L74-L75

This logic correctly accounts for many self-defined body lengths:

https://github.com/netty/netty/blob/e8b5aeddef6f342e5e8ca1e03092c7b278da2915/codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java#L114-L117

However, there are some response types that are not correctly accounted for. This leads to unnecessarily closing a connection and preventing connection reuse in these scenarios.

I have not audited all of the possible response types that may be missing, but the following scenarios are correctly captured in `HttpClientCodec` and `HttpObjectDecoder` but not `HttpServerKeepAliveHandler`:

1. Any response to a HEAD request (which MUST NOT include a body)
2. A successful response to a CONNECT request
3. 304 Not Modified responses

https://github.com/netty/netty/blob/e8b5aeddef6f342e5e8ca1e03092c7b278da2915/codec-http/src/main/java/io/netty/handler/codec/http/HttpClientCodec.java#L265-L303

https://github.com/netty/netty/blob/e8b5aeddef6f342e5e8ca1e03092c7b278da2915/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java#L510-L513

For `HttpServerKeepAliveHandler` to be functionally correct, it should capture these scenarios as well. And perhaps all of the above classes should use some shared logic to guarantee consistency.

**Comment by normanmaurer** [comment]
@Bennett-Lynch good catch... Maybe we should factor out all of this to a helper method and then just re-use ? WDYT ?

**Comment by Bennett-Lynch** [comment]
@normanmaurer I like the idea of some type of helper method, but that method would have to require parameters for both `HttpRequest` and `HttpResponse`, and the `HttpServerKeepAliveHandler` would have to be modified to contain a `Queue<HttpRequest>` or `Queue<HttpMethod>` to support that.

**Comment by normanmaurer** [comment]
Yes
